### DISCUSSION TITLE
Custom writer block separators

### DISCUSSION THREAD

**Discussion by wlupton** [jgm/pandoc#9881] [Q&A] [answered]
Hello. I have a custom writer that, given this input:
```
* A
  - B
  - C

D
```

is generating this output:
```
- A

  - B
  - C

D
```

But I don't want the blank line before the nested list (this is targeted at Typst, and the blank line makes a difference).

I looked into this and it seems that the explanation is that that the default `Writer.Blocks` always separates blocks with a blank line (in this case the `Plain` `A` and the `BulletList` `-B` `-C`. Is this correct?

I modified the writer to use this instead (`cr` is `pandoc.layout.cr` and `concat` is `pandoc.layout.concat`):
```
local blocks = function(blks)
    local function render(blk)
        return concat { Writer.Block(blk), cr }
    end
    return concat(blks:map(render))
end

Writer.Blocks = blocks
```

and I now get this output:
```
- A
  - B
  - C
D
```

Compare with the default Typst writer, which gives:
```
- A
  - B
  - C

D
```

I'd prefer that, but Typst doesn't require the blank line, so my version seems to work.

Finally to my question. What should I really be doing here? Ideally I'd only use standard features of the `pandoc.layout` module and wouldn't need any special tag-specific logic. What I've currently done (unconditionally using `cr` rather than `blankline`) seems rather dangerous.

**Comment by wlupton** [comment]
@tarleb, can you shed any light on this? Thanks.

**Comment by tarleb** [comment, marked as answer]
> I looked into this and it seems that the explanation is that that the default `Writer.Blocks` always separates blocks with a blank line (in this case the `Plain` `A` and the `BulletList` `-B` `-C`. Is this correct?

That's correct.

> What should I really be doing here?

That depends, of course. What the built-in Typst writer is doing is to append blank lines to most elements like BulletList, Para, etc, but not to others, e.g. Plain. The final result is then generated by stacking all rendered blocks on top of each other, but without adding an empty line in-between.

Something like this:

``` lua
local layout = require 'pandoc.layout'

Writer.Block.Para = function (para)
  return {Writer.Inlines(para.content), layout.blankline}
end

Writer.Block.Plain = function (plain)
  return Writer.Inlines(plain.content)
end

Writer.Pandoc = function (doc)
  -- The second argument to `Writer.Blocks` is the separator.
  -- It the separator is optional and defaults to `layout.blankline`.
  return Writer.Blocks(doc.blocks, layout.cr)
end
```

**Comment by wlupton** [comment]
Thanks! I didn't realise about the `Blocks` second argument was the separator. I'll follow your suggestion.

**Reply by tarleb** [reply]
My bad, I forgot to document that feature. It just now pushed a commit to fix that.
